plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+  -- topologically sort these sccs, since we rely on linear (sorted) scoping in our 'genLets' code generation
+  topSortedSccs :: [S.Set PLC.Unique]
+  topSortedSccs =
+    let allLets = M.keysSet rhsTable in
+    -- we remove lambda sccs because we are not interested in them
+    -- for optimization, we just check one of the elements if it is a let unique. lambdas appear only by themselves.
+    filter (\ scc-> S.findMin scc `S.member` allLets)
+    . fmap AMN.vertexSet -- we are not interested in graph structure anymore
+    . fromRight (error "Cycle detected in the depgraph. This shouldn't happen in the first place.") $ AM.topSort $ AM.scc reducedDepGraph
+
+  -- | Groups a given set of lets into one or more multilets and wraps these multilets around a given term.
+  -- The grouping is done through the strongly-connected components
+  -- The input lets is not sorted w.r.t. linear scoping, so this function uses the topological-sort of these SCCs,
+  -- to figure out the correct (dependend/linear) order in which to generate these new multilets.
+  --
+  -- The resulting term is wrapped with linear-scope-sorted LetRecs and LetNonRecs (interspersed between each other because letnonrec and letrec syntax cannot be combined)
+  -- Example: `let {i = e, ...} in let rec {j = e, ...} in let rec {...} in let {...} in ..... in originalTerm`
+  genLets :: S.Set PLC.Unique -- ^ all the let identifiers to wrap around this term
+           -> FloatData -- ^ the remaining data to be floated
+           -> Term tyname name uni a                           -- ^ the term to be wrapped
+           -> State [S.Set PLC.Unique] (Term tyname name uni a)                           -- ^ the final wrapped term
+  genLets lets restDepthTable t = do
+    (hereSccs, restSccs) <- gets $ splitSccs lets
+    put $! restSccs
+    foldM genLetsFromScc t hereSccs
+      where
+        -- | given an SCC, it creates a new (rec or nonrec) let-group from it and wraps it around an accumulated term
+        -- Special case: if the new group and the accumulated term are both letnonrecs,
+        -- it merges them together into a single let-group (i.e. linear scoped).
+        genLetsFromScc :: Term tyname name uni a
+                       -> S.Set PLC.Unique
+                       -> State [S.Set PLC.Unique] (Term tyname name uni a)
+        genLetsFromScc accTerm scc = do
+              res <- forM (S.toList scc) (\ v ->
+                                             case M.lookup v rhsTable of
+                                               Just rhs@(_,_,_,oldDepth) ->
+                                                 -- lift each resulting binding in a list monoid to accumulate them
+                                                 -- visit the generated rhs-term as well for any potential floating
+                                                 (_3.bindingSubterms) (goTerm
```

Comment:
```
rhs & (_3.bindingSubterms) nameThisFunction
```
or whatever the typechecking version of that is.

Nit, but the `value & optic` pattern is pretty common, and I think it's confusing to use optics as functions (even when they work that way).

Reply:
Well, I mean this is still using it as a function but in a conventional way.